### PR TITLE
Allow using ConstraintMatrix::shift if initialized with IndexSet

### PR DESCRIPTION
**Pull request by masterleinad**
For some reason we didn't allow to use `ConstraintMatrix::shift` if the object is initialized with an `ÌndexSet` object. However, the attached test shows that shifting and merging into another `ConstraintMatrix` object is possible in this case, too.

### CODE REVIEW

**Comment by tjhei** [comment]
I am confused and I don't think this works as expected: ``lines_cache`` will be incorrect if an ``IndexSet`` is present, because there should be only an entry for each locally available row.

Secondly, I am confused about this implementation of ``shift`` and would have expected that that the ``IndexSet`` is being shifted too (instead of expecting the indices to be both contained). This should be documented at least.

**Comment by masterleinad** [comment]
@tjhei Thanks, you're totally right! This should now be correct.

File: tests/lac/constraints_shift_01.cc [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  Vector<double> vec(size);
+  for (unsigned int i=0; i<size; ++i)
+    vec(i) = i;
+  constraints1.distribute(vec);
+  vec.print(deallog.get_file_stream(), 3, true, false);
+}
+
+int main()
+{
+  std::ofstream logfile("output");
```
Comment (tjhei) [inline tests/lac/constraints_shift_01.cc]: can you replace these 3 lines by ``initlog();`` please?

File: include/deal.II/lac/constraint_matrix.h [hunk 2]
```diff
@@ -373,6 +373,9 @@ class ConstraintMatrix : public Subscriptor
    * and columns correspond to the same degrees of freedom, you'd generate
    * several constraint objects, then shift them, and finally merge() them
    * together again.
+   *
+   * If this object is initialized with an IndexSet, also the shifted indices
```
Comment (tjhei) [inline include/deal.II/lac/constraint_matrix.h]: "If this object is initialized with an IndexSet, it has to also contain the indices shifted by @p offset."

**Comment by masterleinad** [comment]
At the moment, we only support merging if `local_lines` coincide. That is the main reason why I don't want to modify the `IndexSet` in `ConstraintMatrix::shift`. I agree that shifting the `IndexSet` looks more natural, if we allow to merge `ConstraintMatrix` with different `local_lines`.

**Comment by masterleinad** [comment]
This new version shifts `local_lines` as well.

To be merged after #3789.

**Comment by masterleinad** [comment]
@tjhei Would you mind to have another look?

**Comment by tjhei** [comment]
Looks good to me.

**Pull request merged**